### DISCUSSION TITLE
Executing dags with different versions of same package( linux or python packaage)

### DISCUSSION THREAD

**Discussion by tkansara** [apache/airflow#27099] [Q&A]
HI All
I have a architecture related question for Airflow on K8s. I have airflow running using a standard image, and the webserver and workers have the same image.

If I have a requirement to have a specific linux package or python package for the dags, meaning:

Dag1 needs linux Package with version 1.5
Dag2 needs linux Package with version 1.7 (hypothetical use case)
How can I accomplish this if I am using 1 image of airflow ?

#1) Is it better to use kubernetesPodOPerator for the dag that has out of blue requirement, like a specific package?

#2) Or, would it be better to use k8sPodOPerator for all Dags and install the pythonpackage(like say aws packate) and the required Linux package for each dag at the dag level ?

The disadvantage of #1 is that all package(say python packages) will be installed across all workers, regardless the need.

Any thoughts on this ?

**Comment by potiuk** [comment]
There is a whole chapter about this issue added recently in Airlfow Documentation https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#handling-conflicting-complex-python-dependencies  discussing when you should use what and what are the pros/cons. I think you should read it and you will find answers there. 

Usually there is no "better" - but "it depends" and you should read pros/cons of all solutions and choose what's best for you. 

Note that we do not have "per dag" environment yet that will make top-level parsing works with different dependencies. It is somewhow possible with "standalone DAG processor" already in 2.4* implemented via [AIP-43](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-43+DAG+Processor+separation) but it requires some complexity in your deployment. Better, more complete solution will come when we implement [AIP-46](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-46+Runtime+isolation+for+airflow+tasks+and+dag+parsing ).

Currently the best and simplest you can at most make choose "per TASK" what is the excecution environment of that task (and the best practices doc explain pros/cons of different approaches).
